notebooks/2_feature_engineering.ipynb

```diff
@@ -189,77 +203,134 @@
     "seq2pat.add_constraint(3 <= price.average() <= 4)\n",
     "\n",
     "# Patterns that occur at least twice (A-D)\n",
-    "patterns = seq2pat.get_patterns(min_frequency=2)"
-   ]
+    "patterns = seq2pat.get_patterns(min_frequency=2)\n",
+    "print(\"Frequent Patterns: \", patterns)\n",
+    "\n",
+    "# Encoding of each user with respect to frequent patterns (A-D)\n",
+    "encodings = get_one_hot_encodings(sequences, patterns)\n",
+    "print(\"Encodings:\\n\", encodings)"
+   ],
+   "metadata": {
+    "collapsed": false,
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   }
   },
   {
    "cell_type": "markdown",
-   "metadata": {},
    "source": [
     "* In this example, we have 3 users with certain sequential events, e.g. page visits in-order. \n",
-    "* Notice, how the length of each sequence in the sequence database is different.\n",
-    "* We can mine for frequent patterns in this sequence database while setting a `min_frequency` threshold to denote the minimum number of occurence of a pattern to be considered frequent. \n",
-    "* More importantly, we consider **attributes** that correspond to each sequential event. For example, the price of the item in each page visit. \n",
-    "* Then, we add **constraints** to reason about attributes, here, the average price to be between 3 and 4. \n",
-    "* Pattern mining operates on the sequence database and seeks frequent patterns while satisfying constraints and the minimum frequency threshold. \n",
-    "* See [Seq2Pat Constraints](https://github.com/fidelity/seq2pat/blob/master/notebooks/usage_example.ipynb)."
-   ]
+    "* Notice, how the length of each sequence in the sequence database can be different.\n",
+    "* We can mine for frequent patterns in this sequence database. The `min_frequency` threshold denotes the minimum number of occurrence of a pattern to be considered frequent.\n",
+    "* More importantly, we consider **attributes** that correspond to each sequential event. Here, the price of the item in each page visit.\n",
+    "* Then, we add **constraints** to reason about attributes. Here, the average price to be between 3 and 4.\n",
+    "* Pattern mining operates on the sequence database and seeks frequent patterns that meet the minimum frequency threshold and satisfy constraints.\n",
+    "* See other [Seq2Pat Constraints](https://github.com/fidelity/seq2pat/blob/master/notebooks/usage_example.ipynb) for .\n",
+    "* Finally, we create one-hot encodings of each user with respect to frequent pattern (A-D). The one-hot vector indicates the existence of such pattern in the user's sequence.\n",
+    "* These one-hot encodings of users can be used as input features for the recommendation model.\n",
+    "* In this example, we mined patterns for the entire population. A more advanced approach is to mine patterns that distinguish between the positive vs. negative behavior, e.g., differentiate users who engaged with a certain content from those who do not.\n",
```

Comment:
I reworded this a little bit because it was not clear. I am still on the fence about the DPM part. DPM works with dichotomy, what's the pos/negative here? How would the user know? The only way I can see DPM working for recommenders is to do a positive/negative split _for every content_ separately. Did you have something else in mind?

Reply:
This makes sense. It makes the goal of DPM clearer. I am just thinking perhaps a split for every content seems a little too intensive. It might be a split for a recommended promotion? e.g. subscription/unsubscription to a service after a trial?